okapi-request/lib/OkapiRequest.js

```diff
@@ -272,15 +317,57 @@ class OkapiRequest
   /**
    * eachPromise
    * iterate through an array of items IN SERIES, applying the given async
-   * function to each.
+   * function to each element. If this.streams is non-zero, partition the
+   * array into streams and handling the streams in parallel but each stream
+   * in series.
    * @arg [] arr array of elements
    * @arg function fn function to apply to each element
    * @return promise
    */
   eachPromise(arr, fn)
   {
     if (!Array.isArray(arr)) return Promise.reject(new Error('Array not found'));
-    return arr.reduce((prev, cur) => (prev.then(() => fn(cur))), Promise.resolve());
+
+    /**
+     * eachPromiseSeries
+     * Split an array of Promises into N chunks to be handled in parallel,
+     * the items in each chunk being handled in series. e.g given 100 elements
+     * and a chunk-size of 4, create 4 parallel streams that handle 25 promises
+     * in series, applying the given async function to each.
+     * @arg [] arr array of elements
+     * @arg function fn function to apply to each element
+     * @return promise
+     */
+    const eachPromiseSeries = (arr, fn) => {
```

Comment:
@zburke is there a reason why `eachPromiseSeries` and `eachPromiseParallel` are nested under `eachPromise`?

I'm asking because every time `eachPromise` is executed it will recreate these 2 functions. Could they be moved up to the `OkapiRequest` class?

Reply:
Nope; that's just an artifact of how I did the refactor. Lifting them up a level is a good idea!